2211 do not load unfiltered reports

mkllnk opened the pull request:
#### What? Why?

Closes #2211

Most reports are slow and some even fail to render within the request time. As a first step, we don't load the content of a report until the user has entered some search criteria and submitted the form.

It is still possible to display full reports if the search form is left empty and you click on "Search".

The code also got groomed a little, but I had to stop myself re-factoring to get the job done.

#### What should we test?

All reports. At least, we need to check that every report is not displaying data to start with, but displays data when searching. In addition, the Orders And Distributors report had a few more changes and it would be good to double check that it's still displaying the right thing.

#### Release notes

Reports don't load any data before clicking on "Search" to speed up report access.

#### How is this related to the Spree upgrade?

We are overriding Spree's reports controller. If there is work on the reports controller within the Spree upgrade branch, there could be merge conflicts. But no change here should make the upgrade more difficult, it should just become easier.

sauloperez commented on app/controllers/spree/admin/reports_controller_decorator.rb:
```diff
@@ -53,17 +53,17 @@ def report_types
   end
 
-  # This action is short because we refactored it like bosses
+  # This action is short because we re-factored it like bosses
```
I've always seen `refactor` and never `re-factor` in the computer science literature :thinking:

mkllnk replied:
I've seen both and thought that "re-factor" would be grammatically correct. It was probably the influence of my outdated spell checker that made me do this change. I'm happy to remove it.

mkllnk replied:
Looking at this again, now that we have made more of these methods short, we can probably delete the whole comment.

sauloperez commented on lib/open_food_network/customers_report.rb:
```diff
@@ -1,9 +1,10 @@
+    def initialize(user, params = {}, compile_table = false)
       @params = params
       @user = user
+      @compile_table = compile_table
```
while do you refer to `compile` and not `render`. What's the difference with the BulkCoopReport above?

mkllnk replied:
"render" is used to draw something on the screen. If you you have two `render` statements in a controller, it will raise an error. The intent here is not just to avoid the rendering of data, it's to avoid all the data loading and sorting in the first place.

sauloperez commented on spec/controllers/spree/admin/reports_controller_spec.rb:
```diff
@@ -68,10 +68,13 @@
       it "shows all orders in order cycles I coordinate" do
-        spree_get :orders_and_fulfillment
+        # create test objects
+        [orderA1, orderA2, orderB1, orderB2]
```
why not using `let!` instead of this expression? seems a bit odd to me. The fact that you need a comment feels like a symptom to me.

mkllnk replied:
The thing about the comment is true. I don't know of any pattern for this. If I used `let!`, would need to add a whole block around every single example so that the `let!` happens only for that one example. Or can you use it within `it`?

I would like to create these objects on a case by case basis, individually for each `it`. I could assign them to a variable just for naming purposes, but then we have an unused variable: `test_context = [orderA1, orderA2, orderB1, orderB2]`

How would you do it?

sauloperez commented on spec/controllers/spree/admin/reports_controller_spec.rb:
```diff
@@ -199,74 +211,104 @@
     before { login_as_admin }
 
     it "should build distributors for the current user" do
+      [c1, c2, s1, d1, d2, d3]
```
I would rather create these c1, c2, etc within this "Products & Inventory" context. And if we could provide more meaningful names... it's really hard to understand what this vars are about :pray:

mkllnk replied:
> I would rather create these c1, c2, etc within this "Products & Inventory" context.

At the moment, there are several tests in this context that don't need these objects at all. I can create more contexts. Do you think that would improve readability?

> if we could provide more meaningful names...

Oh yes, totally agree. That would be great. I put that on my list.

sauloperez commented on spec/features/admin/reports_spec.rb:
```diff
@@ -204,28 +204,28 @@
+      expect(page).to     have_content "1512.99" # items total
+      expect(page).to     have_content "1500.45" # taxable items total
+      expect(page).to     have_content "250.08" # sales tax
+      expect(page).to     have_content "20.0" # enterprise fee tax
```
Now that the tabulation is not aligned with the paragraphs above and below I'd simply replace it with a single space.

mkllnk commented on app/controllers/spree/admin/reports_controller_decorator.rb:
```diff
@@ -250,23 +215,29 @@ def xero_invoices
+  # We don't want to render data unless search params are supplied.
+  # Compiling data can take a long time.
+  def render_content?
+    request.post?
```
Yeah, I was wondering about that. My first draft looked for common search parameters. The question is: how difficult is it to maintain this list? Do we look for the presence or absence of search parameters?

One difficulty is that some actions change the `params` object. They set default values. So I need to determine the search state before that. Or I change the code to not change the `params` object, because I think it's bad practice. But Spree's own implementation does that:

https://github.com/spree/spree/blob/baadc8f95cbd25ff8b1741742dbaed75045cbc25/backend/app/controllers/spree/admin/reports_controller.rb#L29

mkllnk commented:
Thank you @sauloperez and @enricostano for your review. Really good feedback. I think I addressed all your concerns in new commits. Let me know what you think.

sauloperez commented on lib/open_food_network/bulk_coop_report.rb:
```diff
@@ -44,25 +46,12 @@ def header
 
     def search
-      permissions.visible_orders.complete.not_state(:canceled).search(params[:q])
+      OpenFoodNetwork::Reports::LineItems.search_orders(permissions, params)
```
No need to specify `OpenFoodNetwork` here. This code is already within that namespace.

mkllnk replied:
You are right, thanks. I changed it.

sauloperez commented on lib/open_food_network/order_and_distributor_report.rb:
```diff
@@ -1,8 +1,12 @@
+      @user = user
+      @render_table = render_table
+
+      @permissions = OpenFoodNetwork::Permissions.new(user)
```
here as well

sauloperez commented:
looks pretty good! Thanks for the effort 👏. This leaves us in a much better position for upcoming refactors/performance improvements.

mkllnk commented on spec/lib/open_food_network/order_and_distributor_report_spec.rb:
```diff
@@ -34,9 +34,9 @@ module OpenFoodNetwork
+	subject = OrderAndDistributorReport.new create(:admin_user), {}, true
 
-        table = subject.table
+        table = subject.send(:line_item_details, [@order])
```
This was just the result of refactoring and I'm not sure how to resolve that. I don't think we should change the method to public just so that we can test it. Strictly speaking, we just want to test the public methods, but I would have needed to rewrite the test for that (more effort). What would you do?

sauloperez replied:
I don't want to make it public either. I wonder if we can still rely on the `#table` method to compare the expected output against.

mkllnk commented:
Okay @sauloperez, I refactored that spec to use the public `table` method. I had to change the created order for that and took the opportunity to clean it up. What do you think now?

I never thought this pull request would grow that big.

sauloperez commented:
Way better now!

daniellemoorhead commented:
@enricostano can you review the changes made by @mkllnk and approve if you're happy please 😄

oeoeaio commented on app/controllers/spree/admin/reports_controller_decorator.rb:
```diff
@@ -250,23 +215,53 @@ def xero_invoices
+      # `button` is included in all forms. It's not important for searching,
+      # but the Users & Enterprises report doesn't have any other parameter
+      # for an empty search. So we use this one to display data.
+      :button,
```
Could we just use the presence of a `:button` param as a proxy for a search request for all reports? That would seem to negate the need to check all of these other params as well...?

oeoeaio commented on app/controllers/spree/admin/reports_controller_decorator.rb:
```diff
@@ -289,42 +284,42 @@ def load_data
+      ).html_safe
+    end
+    { name: name, description: description }
   end
```
Nice. 👌

sstead commented:
**Testing note:**
- yes reports initially show no data, then the data loads when you hit Search.
- The same is true in superadmin- except for one report (sales totals) but I don't think this is a problem.

- Regarding the 'orders and distributors' report - it's behaving the same as on production. But there seems to be an exisitng issue with one column (shipping cost) not displaying anything.

https://docs.google.com/document/d/1lwevBhGMqh0WC9waDysGG4ZuaVd4efCzt_ItL4JWSSI/edit#
